### PR TITLE
Avoid memory leaks

### LINKED ISSUE #45: Memory leaks when running multiple threads
There seems to be an issue with the underlying `threadpool` crate in that memory is not freed after testing. This is not great as it is common to run tests under valgrind to detect leaks while shooting for full coverage.

### PR DESCRIPTION
**Pull request by Felix-El**
This commit replaces the `threadpool` crate with a handcrafted solution based on scoped threads. This leaves `valgrind` much happier than before. We also lose some dependency baggage.

Closes #45

### CODE REVIEW

File: src/pool.rs
Comment (hanna-kruppe) [inline src/pool.rs]: This is wrong: one worker thread will acquire the lock here and run all the tasks sequentially, while the other worker threads will be blocked until the end and then see an empty iterator. At least I'm 99% sure that's what will happen -- if there is a subtle reason why it actually works, I'd advocate for making the code less subtle!

The mechanical alternative is to only acquire the lock for the duration of each `next()` call. This requires desugaring the for-loop a bit. Perhaps extract the "get next task" bit into a dedicated closure and write the workers as `while let Some(task) = next_task() { task(); }`?

File: src/pool.rs
Comment (hanna-kruppe) [inline src/pool.rs]: Since this type is only ever used as `ScopedPool(vec_of_tasks).run(num_threads)` I would suggest deleting this type and the `<I>` type parameters. A single function without generics would be sufficient and more straightforward: `fn run_all(tasks: Vec<BoxedTask>, num_threads: Option<usize>)`.

In general this is a stylistic question that people can and will disagree with, but in this specific case I think the extra abstraction obscures what's actually going on. For example, it's useful to know that advancing the iterator doesn't do any significant work on its own (otherwise holding a lock while advancing it may be problematic) and that's not true for arbitrary `IntoIterator`s, but it's true for `Vec`.

**Review by hanna-kruppe — COMMENTED** [review]
I started looking into what it would take to replace the threadpool dependency, for overlapping reasons, and then I saw this PR. So, first of all, thank you for writing the patch! I have zero authority to accept or reject it, but I'd also like to see this change done. However, I think the current version has a bug, it doesn't seem to actually (meaningfully) use multiple threads. And while I was at it, I made another suggestion (but again, my opinion carries no weight here).

File: src/pool.rs
Comment (Felix-El) [reply src/pool.rs]: I agree. Updated to what you propose here.

File: src/pool.rs
Comment (Felix-El) [reply src/pool.rs]: Right, there is no real need for a structure so I changed to a `fn`. However I left the `IntoIterator` for another review round, as I do see a valid reason for it to be generic. See the newly-added tests where we create dummy-tasks on-demand rather than pre-creating them to a Vec. True, we expect the operation to be cheap enough.

**Comment by Felix-El** [comment]
@hanna-kruppe, many thanks you for your review. Appreciate you took the time to look in-depth.
I've refactored my approach a couple of times before my initial commit. Actually I've been there before, at roughly what you proposed above. Anyway, I hope the update resonates better with you and an authority.

File: src/pool.rs
Comment (Felix-El) [reply src/pool.rs]: If it really makes a difference, I could add a `pub(crate) fn` version which accepts `Vec` and delegates to the generic version which becomes private.

Comment (hanna-kruppe) [reply src/pool.rs]: I'm not sold on those tests (and thus, on keeping a more complicated interface for their sake). Recording thread IDs of everyone who tries to get a task from the iterator is a clever hack, but it *is* a hack, not a test of behavior that actually matters. The previous buggy version of this patch would have passed those tests, since it did call `next()` on the tasks iterator from every thread at least once! A hypothetical buggy version that pulled tasks from the iterator but didn't run them would also pass the test. Conversely, the tests nail down current implementation details and would unnecessarily fail if they changed (e.g., if the main thread consumed the iterator itself and pushed all tasks to a multi-consumer channel that the workers pull from). It's just really hard to have good unit tests for thread pools, especially for non-functional requirements such as "actually utilize all the threads".

File: src/lib.rs
Comment (LukasKalbertodt) [inline src/lib.rs:514]:
```suggestion
        for test in tests {
```

Should work?

File: src/lib.rs
Comment (LukasKalbertodt) [inline src/lib.rs:505]: Can you move the `let num_tests` line to its previous position (right above the loop) to reduce the diff size?

File: src/lib.rs
Comment (LukasKalbertodt) [inline src/lib.rs:512]: Shouldn't this work?

```suggestion
        let mut tasks = Vec::new();
```

File: src/pool.rs
Comment (LukasKalbertodt) [inline src/pool.rs:4]: I don't think these type aliases pull their weight. With my suggestion from a previous comment, they are only used in one spot. So please remove the aliases and just write out the type in the `scoped_run_tasks` function.

File: src/lib.rs
Comment (LukasKalbertodt) [inline src/lib.rs:531]: The function is pretty short and only used in a single place (here), so why not inline it? Similar to what @hanna-kruppe said above, I think in this case we lose more by obscuring what actually happens compared to what we gain by black box abstraction. 

Further, using `thread::scope` is actually problematic when looking at the next loop: that expects to run while the worker threads are doing something, in order to print something immediately after a test is finished. `thread::scope` waits for all threads to be done before returning. So in your code, all finished tests would fill up the channel, and only after all tests are done does the following loop run and print all tests. 
This can be fixed by moving the printing loop below into the `thread::scope` closure, which also requires inlining the function.

Aaand in that case we can also get rid of the first loop, since we already have a vector of "tasks" (the trials) we can iterate over.

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
Thanks for your PR! Looking at the (lack of) activity of the `threadpool` crate, it does seem like a good idea to replace the dependency. Especially since doing it ourselves is really not hard. 
Thank you @hanna-kruppe for already reviewing this before. Out of curiosity: what made you look into removing the dependency?

I have a few problems with the proposed changes, however. I left a few inline comments with my reasoning. The main functional problem is that the outcomes are only printed after all tests have passed. 

However, to avoid another contributer-maintainer roundtrip (which, due to mainly me, tends to take a while), I just applied all changes already and rebased your PR as well. That way I can release this change soon.

**Comment by LukasKalbertodt** [comment]
Due to me not being able to push to your branch, I had to merge manually, which (paired with a rebase), github does not recognize. But here, your changes are merged: https://github.com/LukasKalbertodt/libtest-mimic/commit/d194ff99f600e9adecd1ffb262e764b9ab8e03d2

**Comment by hanna-kruppe** [comment]
> Thank you @hanna-kruppe for already reviewing this before. Out of curiosity: what made you look into removing the dependency?

Every once in a while I like the exercise of [reading through the lockfile](https://matklad.github.io/2021/09/04/fast-rust-builds.html#Read-The-Lockfile). I was surprised to see `num_cpus` in one of my lockfiles, since it's a relic of the bad old days before `std::thread::available_parallelism`. After tracing that back to `threadpool` and looking at its lack of activity, open issues, and skimming the implementation, I thought there's gotta be a simpler way to run some tests in parallel.

**Comment by LukasKalbertodt** [comment]
@hanna-kruppe Ah, that makes. I also do regularly do that (mostly via `cargo tree`). I'm simply not personally using `libtest-mimic`, so I tend to not notice this stuff. 

And by the way (for the two of you): I just released v0.8.0 with these changes.

File: src/lib.rs
Comment (hanna-kruppe) [reply src/lib.rs:531]: I think the way you inlined it in d194ff99f600e9adecd1ffb262e764b9ab8e03d2 has a bug: it spawns `num_threads` scoped threads, each of which only takes *one* trial from `iter` and then terminates. So if you have fewer threads than trials, not all tests are executed.

Comment (LukasKalbertodt) [reply src/lib.rs:531]: Oh no

Comment (LukasKalbertodt) [reply src/lib.rs:531]: Ooof, that's awkward. Thanks for catching it! Should be fixed in d6ac84fe44f15bddf801a0a59599995b9363068c. Would you mind reviewing that, to avoid me publishing another avoidable bug?

Comment (hanna-kruppe) [reply src/lib.rs:531]: Left a comment on that commit.

Comment (LukasKalbertodt) [reply src/lib.rs:531]: Thanks for reviewing. I removed the outdated comment in e17efb916a68573063449d40ea368a463703e991 and also added additional tests in 41a18561074fab4602505b5b646fd352ea04ac66 to catch this kind of bug in the future. Thanks again for catching this, will release momentarily.

**Pull request closed without merging**
